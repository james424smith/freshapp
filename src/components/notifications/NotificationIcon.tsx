import React, { useState } from "react";
import { useSelector } from "react-redux";
import {} from "../../redux/actions";
import IconBadge from "react-native-icon-badge";
import Colors from "../../constants/styles/colors";
import Text from "../StyledText";
import styles from "../../styles/NotificationsScreenStyles";
import { Animated } from "react-native";
import { IRootReducerType } from "../../redux/reducers";
import Notifications from "../../../assets/icons/bottom_navigation/005-bell.svg";
import { useTheme } from "@react-navigation/native";

interface Props {
  mainIconSize: number;
  focused: boolean;
}

const NotificationIcon = (props: Props) => {
  const [animationValue] = useState<Animated.Value>(new Animated.Value(0));
  const theme = useTheme();
  const { mainIconSize, focused } = props;
  const unreadCount = useSelector<IRootReducerType, number>(
    ({ notificationsDetails }) =>
      notificationsDetails.notificationsDetails.unreadCount
  );

  const rotateInterpolate = animationValue.interpolate({
    inputRange: [0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8],
    outputRange: [
      "0deg",
      "-30deg",
      "-60deg",
      "-30deg",
      "0deg",
      "30deg",
      "60deg",
      "30deg",
      "0deg",
    ],
  });
  const animatedStyle = { transform: [{ rotate: rotateInterpolate }] };
  return (
    <Animated.View style={animatedStyle}>
      <IconBadge
        MainElement={
          <Notifications
            testID={"notification-icon-button"}
            width={mainIconSize}
            height={mainIconSize}
            fill={focused ? theme.colors.tabNavigatorBlueToDark : Colors.grey}
          />
        }
        BadgeElement={
          <Text style={styles.badgeElementStyle}>{unreadCount}</Text>
        }
        IconBadgeStyle={styles.iconBadgeStyle}
        Hidden={!unreadCount}
      />
    </Animated.View>
  );
};

export default NotificationIcon;
