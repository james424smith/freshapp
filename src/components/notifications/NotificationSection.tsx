import React from "react";
import { View, TouchableOpacity } from "react-native";
import Text from "../StyledText";
import styles from "../../styles/NotificationsScreenStyles";
import {
  NotificationDetails,
  ReadNotification,
} from "../../interfaces/notificationsTypes";
import {
  ASSIGNMENT,
  PAYSLIP,
  FLIGHT,
  PERSONAL,
  NEWSLETTER,
  GENERAL,
} from "../../constants/notificationsConstants";
import { useTheme, Theme } from "@react-navigation/native";
import valueOrDefault from "../../common/valueOrDefault";
import * as HandleLinkHelper from "../../common/handleLink";

interface Props {
  markAsRead: (payload: ReadNotification) => void;
  notificationsList?: NotificationDetails[];
}

async function handleOnClick(
  notification: NotificationDetails,
  markAsRead: (payload: ReadNotification) => void
) {
  if (notification.isRead === false) {
    const payload: ReadNotification = {
      notificationRead: true,
      notificationId: notification.notificationId.toString(),
    };
    markAsRead(payload);
  }

  switch (notification.category) {
    case ASSIGNMENT:
      HandleLinkHelper.handleLink("crewcompanion", "assignment");
      break;
    case PAYSLIP:
      HandleLinkHelper.handleLink("crewcompanion", "salaries");
      break;
    case FLIGHT:
      HandleLinkHelper.handleLink("crewcompanion", "flights");
      break;
    case PERSONAL:
      HandleLinkHelper.handleLink("crewcompanion", "documents");
      break;
    case NEWSLETTER:
      HandleLinkHelper.handleLink("crewcompanion", "news");
      break;
    case GENERAL:
      break;
    default:
  }
}

function renderNotification(
  theme: Theme,
  item: NotificationDetails,
  key: number,
  markAsRead: (payload: ReadNotification) => void
) {
  return (
    <View key={key}>
      <View style={styles.notificationContainer}>
        <TouchableOpacity
          testID={"notification-button"}
          onPress={() => handleOnClick(item, markAsRead)}
        >
          <Text
            style={
              item.isRead
                ? { ...styles.dateReadStyle, color: theme.colors.text }
                : {
                    ...styles.dateUnreadStyle,
                    color: theme.colors.createNewAccountColorText,
                  }
            }
          >
            {valueOrDefault(item.date, "")}
          </Text>

          <Text
            style={
              item.isRead
                ? { ...styles.description, color: theme.colors.text }
                : {
                    ...styles.descriptionBold,
                    color: theme.colors.createNewAccountColorText,
                  }
            }
          >
            {valueOrDefault(item.description, "")}
          </Text>
        </TouchableOpacity>
      </View>
      <View style={styles.lineStyle} />
    </View>
  );
}

const NotificationSection = (props: Props) => {
  const { notificationsList, markAsRead } = props;
  const theme = useTheme();

  return (
    <View>
      <View style={styles.notificationsView}>
        {(valueOrDefault(
          notificationsList,
          []
        ) as NotificationDetails[]).map(
          (item: NotificationDetails, i: number) =>
            renderNotification(theme, item, i, markAsRead)
        )}
      </View>
    </View>
  );
};

export default NotificationSection;
