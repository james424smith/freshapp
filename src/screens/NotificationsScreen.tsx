import React, { useEffect } from "react";
import { View, ScrollView, RefreshControl } from "react-native";
import { useSelector, useDispatch } from "react-redux";
import {
  getNotificationsDetails,
  sendNotificationsDetails,
  setRefreshingNotificationsDetails,
} from "../redux/actions";
import { Notification, ReadNotification } from "../interfaces";
import styles from "../styles/NotificationsScreenStyles";
import NotificationsSection from "../components/notifications/NotificationSection";

import { IRootReducerType } from "../redux/reducers";
import { useNavigation } from "@react-navigation/native";

const NotificationsScreen = () => {
  const notifications = useSelector<IRootReducerType, Notification>(
    ({ notificationsDetails }) => notificationsDetails.notificationsDetails
  );
  const isLoading = useSelector<IRootReducerType, boolean>(
    ({ notificationsDetails }) => notificationsDetails.loader
  );
  const dispatch = useDispatch();
  const navigation = useNavigation();

  const markAsRead = (payload: ReadNotification) => {
    dispatch(sendNotificationsDetails(payload));
  };

  useEffect(() => {
    const _navListener = navigation.addListener("focus", () => {
      dispatch(getNotificationsDetails());
    });
    return () => {
      _navListener();
    };
  }, [dispatch, navigation]);

  const refreshData = () => {
    dispatch(setRefreshingNotificationsDetails(true));
    dispatch(getNotificationsDetails());
  };

  return (
    <ScrollView
      testID={"scrollview"}
      style={styles.scrollViewStyle}
      refreshControl={
        <RefreshControl refreshing={isLoading} onRefresh={refreshData} />
      }
    >
      <View>
        <NotificationsSection
          notificationsList={notifications?.notifications}
          markAsRead={markAsRead}
        />
      </View>
    </ScrollView>
  );
};

export default NotificationsScreen;
