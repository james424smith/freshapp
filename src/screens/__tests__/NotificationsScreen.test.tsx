import React from "react";
import {
  render,
  cleanup,
  waitFor,
  fireEvent,
} from "@testing-library/react-native";
import NotificationsScreen from "../NotificationsScreen";
import notifications from "../../../__mocks__/fake-data/notifications.json";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";

describe("NotificationsScreen snapshot test", () => {
  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
    jest.restoreAllMocks();
  });

  it("should render NotificationsScreen when loader is false and photoSmall is defined and hasBoth is true", async () => {
    const mockedState = {
      notificationsDetails: {
        notificationsDetails: notifications,
        loader: false,
      },
      network: { isConnected: true },
    };
    const mockedStore = configureMockStore();

    const { toJSON, getAllByTestId, getByTestId } = render(
      <Provider store={mockedStore(mockedState)}>
        <NotificationsScreen />
      </Provider>
    );
    fireEvent.press(getAllByTestId("notification-button")[0]);

    getByTestId("scrollview").props.refreshControl.props.onRefresh();
    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });
  it("should render NotificationsScreen when loader is false and photoSmall is defined and hasBoth is false", async () => {
    jest.useRealTimers();
    const mockedState = {
      notificationsDetails: {
        notificationsDetails: notifications,
        loader: true,
      },
      network: { isConnected: true },
    };
    const mockedStore = configureMockStore();

    const { toJSON } = render(
      <Provider store={mockedStore(mockedState)}>
        <NotificationsScreen />
      </Provider>
    );

    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });
});
