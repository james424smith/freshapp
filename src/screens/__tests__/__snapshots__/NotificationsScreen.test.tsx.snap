// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationsScreen snapshot test should render NotificationsScreen when loader is false and photoSmall is defined and hasBoth is false 1`] = `
<RCTScrollView
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={true}
    />
  }
  style={
    Object {
      "flexGrow": 1,
    }
  }
  testID="scrollview"
>
  <RCTRefreshControl />
  <View>
    <View>
      <View>
        <View
          style={
            Object {
              "paddingTop": 45,
            }
          }
        >
          <View>
            <View
              style={
                Object {
                  "marginHorizontal": 60,
                }
              }
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
                testID="notification-button"
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#2A2E43",
                        "fontFamily": "Roboto-Regular",
                      },
                      Object {
                        "color": undefined,
                        "fontSize": 22.5,
                        "fontWeight": "bold",
                        "paddingBottom": 11.5,
                      },
                    ]
                  }
                >
                  02 Jul 2019, 17:14
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#2A2E43",
                        "fontFamily": "Roboto-Regular",
                      },
                      Object {
                        "color": undefined,
                        "fontSize": 26.5,
                        "fontWeight": "bold",
                      },
                    ]
                  }
                >
                  You have expired/expiring documentation. Please check.
                </Text>
              </View>
            </View>
            <View
              style={
                Object {
                  "borderBottomColor": "#78849E",
                  "borderBottomWidth": 0.2,
                  "marginBottom": 45,
                  "marginLeft": 45,
                  "marginTop": 45,
                }
              }
            />
          </View>
          <View>
            <View
              style={
                Object {
                  "marginHorizontal": 60,
                }
              }
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
                testID="notification-button"
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#2A2E43",
                        "fontFamily": "Roboto-Regular",
                      },
                      Object {
                        "color": undefined,
                        "fontSize": 22.5,
                        "fontWeight": "bold",
                        "paddingBottom": 11.5,
                      },
                    ]
                  }
                >
                  02 Jul 2019, 17:07
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#2A2E43",
                        "fontFamily": "Roboto-Regular",
                      },
                      Object {
                        "color": undefined,
                        "fontSize": 26.5,
                        "fontWeight": "bold",
                      },
                    ]
                  }
                >
                  You have expired/expiring documentation. Please check.
                </Text>
              </View>
            </View>
            <View
              style={
                Object {
                  "borderBottomColor": "#78849E",
                  "borderBottomWidth": 0.2,
                  "marginBottom": 45,
                  "marginLeft": 45,
                  "marginTop": 45,
                }
              }
            />
          </View>
          <View>
            <View
              style={
                Object {
                  "marginHorizontal": 60,
                }
              }
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
                testID="notification-button"
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#2A2E43",
                        "fontFamily": "Roboto-Regular",
                      },
                      Object {
                        "color": undefined,
                        "fontSize": 22.5,
                        "fontWeight": "bold",
                        "paddingBottom": 11.5,
                      },
                    ]
                  }
                >
                  02 Jul 2019, 17:04
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#2A2E43",
                        "fontFamily": "Roboto-Regular",
                      },
                      Object {
                        "color": undefined,
                        "fontSize": 26.5,
                        "fontWeight": "bold",
                      },
                    ]
                  }
                >
                  You have expired/expiring documentation. Please check.
                </Text>
              </View>
            </View>
            <View
              style={
                Object {
                  "borderBottomColor": "#78849E",
                  "borderBottomWidth": 0.2,
                  "marginBottom": 45,
                  "marginLeft": 45,
                  "marginTop": 45,
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`NotificationsScreen snapshot test should render NotificationsScreen when loader is false and photoSmall is defined and hasBoth is true 1`] = `
<RCTScrollView
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={false}
    />
  }
  style={
    Object {
      "flexGrow": 1,
    }
  }
  testID="scrollview"
>
  <RCTRefreshControl />
  <View>
    <View>
      <View>
        <View
          style={
            Object {
              "paddingTop": 45,
            }
          }
        >
          <View>
            <View
              style={
                Object {
                  "marginHorizontal": 60,
                }
              }
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
                testID="notification-button"
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#2A2E43",
                        "fontFamily": "Roboto-Regular",
                      },
                      Object {
                        "color": undefined,
                        "fontSize": 22.5,
                        "fontWeight": "bold",
                        "paddingBottom": 11.5,
                      },
                    ]
                  }
                >
                  02 Jul 2019, 17:14
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#2A2E43",
                        "fontFamily": "Roboto-Regular",
                      },
                      Object {
                        "color": undefined,
                        "fontSize": 26.5,
                        "fontWeight": "bold",
                      },
                    ]
                  }
                >
                  You have expired/expiring documentation. Please check.
                </Text>
              </View>
            </View>
            <View
              style={
                Object {
                  "borderBottomColor": "#78849E",
                  "borderBottomWidth": 0.2,
                  "marginBottom": 45,
                  "marginLeft": 45,
                  "marginTop": 45,
                }
              }
            />
          </View>
          <View>
            <View
              style={
                Object {
                  "marginHorizontal": 60,
                }
              }
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
                testID="notification-button"
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#2A2E43",
                        "fontFamily": "Roboto-Regular",
                      },
                      Object {
                        "color": undefined,
                        "fontSize": 22.5,
                        "fontWeight": "bold",
                        "paddingBottom": 11.5,
                      },
                    ]
                  }
                >
                  02 Jul 2019, 17:07
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#2A2E43",
                        "fontFamily": "Roboto-Regular",
                      },
                      Object {
                        "color": undefined,
                        "fontSize": 26.5,
                        "fontWeight": "bold",
                      },
                    ]
                  }
                >
                  You have expired/expiring documentation. Please check.
                </Text>
              </View>
            </View>
            <View
              style={
                Object {
                  "borderBottomColor": "#78849E",
                  "borderBottomWidth": 0.2,
                  "marginBottom": 45,
                  "marginLeft": 45,
                  "marginTop": 45,
                }
              }
            />
          </View>
          <View>
            <View
              style={
                Object {
                  "marginHorizontal": 60,
                }
              }
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
                testID="notification-button"
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#2A2E43",
                        "fontFamily": "Roboto-Regular",
                      },
                      Object {
                        "color": undefined,
                        "fontSize": 22.5,
                        "fontWeight": "bold",
                        "paddingBottom": 11.5,
                      },
                    ]
                  }
                >
                  02 Jul 2019, 17:04
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#2A2E43",
                        "fontFamily": "Roboto-Regular",
                      },
                      Object {
                        "color": undefined,
                        "fontSize": 26.5,
                        "fontWeight": "bold",
                      },
                    ]
                  }
                >
                  You have expired/expiring documentation. Please check.
                </Text>
              </View>
            </View>
            <View
              style={
                Object {
                  "borderBottomColor": "#78849E",
                  "borderBottomWidth": 0.2,
                  "marginBottom": 45,
                  "marginLeft": 45,
                  "marginTop": 45,
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
